proxy/search/async.go

```diff
@@ -66,100 +71,217 @@ func (si *Ingestor) StartAsyncSearch(ctx context.Context, r AsyncRequest) (Async
-	if si.config.HotReadStores != nil && len(si.config.HotReadStores.Shards) > 0 {
-		searchStores = si.config.HotReadStores
+func (si *Ingestor) FetchAsyncSearchResult(ctx context.Context, r FetchAsyncSearchResultRequest) (FetchAsyncSearchResultResponse, DocsIterator, error) {
+	// TODO: should we support QueryWantsOldData?
+	searchStores, err := si.getAsyncSearchStores()
+	if err != nil {
+		return FetchAsyncSearchResultResponse{}, nil, err
 	}
 
 	req := storeapi.FetchAsyncSearchResultRequest{
 		SearchId: r.ID,
-		WithDocs: r.WithDocs,
 		Size:     int32(r.Size),
 		Offset:   int32(r.Offset),
 	}
 
-	done := true
-	var expiration time.Time
-	var aggQueries []seq.AggregateArgs
+	fracsDone := 0
+	fracsInQueue := 0
+	histInterval := seq.MID(0)
+	pr := FetchAsyncSearchResultResponse{}
+	mergeStoreResp := func(sr *storeapi.FetchAsyncSearchResultResponse, replica string) {
+		pr.DiskUsage += sr.DiskUsage
+		fracsInQueue += int(sr.FracsQueue)
+		fracsDone += int(sr.FracsDone)
+
+		histInterval = seq.MID(sr.HistogramInterval)
+
+		ss := sr.Status.MustAsyncSearchStatus()
+		pr.Status = mergeAsyncSearchStatus(pr.Status, ss)
+
+		for _, errStr := range sr.GetResponse().GetErrors() {
+			pr.QPR.Errors = append(pr.QPR.Errors, seq.ErrorSource{
+				ErrStr: errStr,
+				Source: si.sourceByClient[replica],
+			})
+		}
+
+		t := sr.ExpiresAt.AsTime()
+		if pr.ExpiresAt.IsZero() || pr.ExpiresAt.After(t) {
+			pr.ExpiresAt = t
+		}
+		t = sr.StartedAt.AsTime()
+		if pr.StartedAt.IsZero() || pr.StartedAt.After(t) {
+			pr.StartedAt = t
+		}
+
+		qpr := responseToQPR(sr.Response, si.sourceByClient[replica], false)
+		seq.MergeQPRs(&pr.QPR, []*seq.QPR{qpr}, r.Size+r.Offset, histInterval, r.Order)
+	}
 
-	var qprs []*seq.QPR
+	var aggQueries []seq.AggregateArgs
 	anyResponse := false
-	histInterval := seq.MID(0)
-	aggsCount := 0
-	order := seq.DocsOrderAsc
 	for _, shard := range searchStores.Shards {
-		var storeResp *storeapi.FetchAsyncSearchResultResponse
-		var err error
-		var replica string
-		for _, replica = range shard {
-			storeResp, err = si.clients[replica].FetchAsyncSearchResult(ctx, &req)
+		for _, replica := range shard {
+			storeResp, err := si.clients[replica].FetchAsyncSearchResult(ctx, &req)
 			if err != nil {
 				if status.Code(err) == codes.NotFound {
 					continue
 				}
-				return FetchAsyncSearchResultResponse{}, err
+				return FetchAsyncSearchResultResponse{}, nil, err
+			}
+			anyResponse = true
+			mergeStoreResp(storeResp, replica)
+			if len(aggQueries) == 0 {
+				for _, agg := range storeResp.Aggs {
+					aggQueries = append(aggQueries, seq.AggregateArgs{
+						Func:      agg.Func.MustAggFunc(),
+						Quantiles: agg.Quantiles,
+					})
+				}
 			}
 			break
 		}
+	}
+	if !anyResponse {
+		return FetchAsyncSearchResultResponse{}, nil, status.Error(codes.NotFound, "async search result not found")
+	}
+
+	if fracsDone != 0 {
+		pr.Progress = float64(fracsDone+fracsInQueue) / float64(fracsDone)
```

Comment:
should not it be the opposite?
```suggestion
		pr.Progress = float64(fracsDone) / float64(fracsDone+fracsInQueue)
```